extensions/ql-vscode/src/common/qlpack-language.ts

```diff
@@ -0,0 +1,30 @@
+import { load } from "js-yaml";
+import { readFile } from "fs-extra";
+import { QlPackFile } from "../packaging/qlpack-file";
+import { QueryLanguage } from "./query-language";
+
+/**
+ * @param qlpackPath The path to the `qlpack.yml` or `codeql-pack.yml` file.
+ * @return the language of the given qlpack file, or undefined if the file is
+ * not a valid qlpack file or does not contain exactly one language.
+ */
+export async function getQlPackLanguage(
+  qlpackPath: string,
+): Promise<QueryLanguage | undefined> {
+  const qlPack = load(await readFile(qlpackPath, "utf8")) as
+    | QlPackFile
+    | undefined;
```

Comment:
I believe this function as it's currently written is safe, but also this type isn't entirely correct. As far as I can tell the `load` function can return anything, and if it can't parse the file contents as yaml then it returns a string of the entire file contents. It would be safer to use `unknown` and then have a type checking function to determine if the object is a qlpack object.

The danger of the current code is the manual checking of the `dependencies` object and not checking any of the other fields. If someone in the future does try to use on the other fields it could cause an error.

What I'm wondering about is whether it's worth the extra complexity to make it safer. I'm tempted to say yes, we should do everything we can to make the types as accurate as possible and avoid any unexpected behaviour. What do you think?

Reply:
Yes, I agree. The easiest way to do this is just to generate a schema from the `QlpackFile` type and use that to validate the returned YAML, so I'll add that.